File: packages/platools-js/src/cli/doctor.ts
/**
 * `platools doctor` CLI command.
 *
 * Ported from `platools/cli/doctor.py`. Mirrors the Python flag
 * surface: positional `module`, `--json`. Module loading uses
 * dynamic ESM `import()` so consumers can point at their own
 * barrel file (`platools doctor ./dist/tools.js`, for example).
 *
 * Exit code:
 *   - 0 if there are no error-severity findings
 *   - 1 if any error finding is present (CI-friendly)
 *   - 2 on argument / file errors
 */

import { pathToFileURL } from "node:url";
import { resolve as resolvePath } from "node:path";

import { Platools } from "../platools.js";
import { ToolRegistry } from "../core/registry.js";
import { analyzeRegistry, analyzeTools } from "../doctor/analyzer.js";
import { formatReport, reportToJson } from "../doctor/reporter.js";
import type { DoctorReport } from "../doctor/types.js";
import type { ToolDef } from "../types.js";

export interface DoctorCliOptions {
  readonly modulePath?: string;
  readonly outputJson?: boolean;
  readonly stdout?: (line: string) => void;
  readonly stderr?: (line: string) => void;
}

/**
 * Build a unified `ToolRegistry` from a user-provided module path.
 *
 * SDK consumers structure their app like this:
 *
 *     // myApp/tools.ts
 *     import { Platools } from "@platools/sdk";
 *     export const platools = new Platools();
 *     platools.tool({...}, async (...) => ...);
 *
 * `platools doctor ./dist/tools.js` dynamically imports the
 * module, finds every exported `Platools` instance, and merges
 * their registries into a fresh composite registry the analyzer
 * walks. Empty modules surface as "Tools: 0 registered".
 */
export async function loadRegistryFromModule(
  modulePath: string | undefined,
): Promise<ToolRegistry> {
  const composite = new ToolRegistry();
  if (modulePath === undefined || modulePath === "") return composite;

  // Accept both bare specifiers (e.g. `@myorg/tools`) and relative
  // file paths. File paths are normalized via `pathToFileURL` so
  // the ESM loader gets a `file://` URL on all platforms.
  const spec = modulePath.startsWith(".") || modulePath.includes("/") || modulePath.endsWith(".js") || modulePath.endsWith(".mjs")
    ? pathToFileURL(resolvePath(process.cwd(), modulePath)).href
    : modulePath;

  const module: Record<string, unknown> = (await import(spec)) as Record<string, unknown>;
  for (const value of Object.values(module)) {
    if (value instanceof Platools) {
      for (const tool of value.registry.all()) {
        if (!composite.has(tool.name)) composite.register(tool);
      }
    }
  }
  return composite;
}

/**
 * Programmatic entry point for `platools doctor`. Returns the
 * process exit code so the CLI bin can `process.exit(code)`.
 */
export async function runDoctor(options: DoctorCliOptions = {}): Promise<number> {
  const stdout = options.stdout ?? ((line: string) => process.stdout.write(line));
  const stderr = options.stderr ?? ((line: string) => process.stderr.write(line));

  let registry: ToolRegistry;
  try {
    registry = await loadRegistryFromModule(options.modulePath);
  } catch (err) {
    stderr(`platools doctor: failed to load module: ${formatError(err)}\n`);
    return 2;
  }

  const report = analyzeRegistry(registry);
  if (options.outputJson === true) {
    stdout(`${reportToJson(report)}\n`);
  } else {
    stdout(formatReport(report));
  }
  return report.hasErrors() ? 1 : 0;
}

/** Convenience helper used by tests — analyze a pre-loaded list. */
export function runDoctorOnTools(tools: readonly ToolDef[]): DoctorReport {
  return analyzeTools(tools);
}

export async function doctorCommand(argv: readonly string[]): Promise<number> {
  const { modulePath, outputJson, help } = parseDoctorArgs(argv);
  if (help) {
    process.stdout.write(doctorHelpText());
    return 0;
  }
  return runDoctor({ modulePath, outputJson });
}

export function doctorHelpText(): string {
  return [
    "Usage: platools doctor [module] [--json]",
    "",
    "Static tool-graph analyzer for Platools.",
    "",
    "Positional:",
    "  module       Optional path/specifier to a module exporting a Platools",
    "               instance (e.g. ./dist/tools.js).",
    "",
    "Options:",
    "  --json       Emit a machine-readable JSON report.",
    "  -h, --help   Show this help and exit.",
    "",
  ].join("\n");
}

export function parseDoctorArgs(argv: readonly string[]): {
  modulePath?: string;
  outputJson: boolean;
  help: boolean;
} {
  let modulePath: string | undefined;
  let outputJson = false;
  let help = false;
  for (const arg of argv) {
    if (arg === "--json") outputJson = true;
    else if (arg === "-h" || arg === "--help") help = true;
    else if (!arg.startsWith("-")) modulePath = arg;
  }
  return modulePath === undefined
    ? { outputJson, help }
    : { modulePath, outputJson, help };
}

function formatError(err: unknown): string {
  if (err instanceof Error) return err.message;
  return String(err);
}

File: packages/platools-py/platools/cli/doctor.py
"""`platools doctor` CLI command — PLATOS-18.

Usage:

    platools doctor                       # local introspection
    platools doctor my_app.tools          # explicit module to load
    platools doctor --json                # machine-readable output

Loads tool definitions either from an explicit Python module path
(`my_app.tools` — same convention pytest uses for `--rootdir`) or
from the calling-process registry. The platform-fetch path
(`--platform-url`) is documented in the spec but lands with the
admin-tooling task; the local path is the ship gate Phase D needs.

Exit code:
  - 0 if there are no `error`-severity findings
  - 1 if any error finding is present (CI-friendly)
"""

from __future__ import annotations

import argparse
import importlib
import json
import sys
from collections.abc import Sequence
from typing import TextIO

from platools.core.registry import ToolRegistry
from platools.doctor import analyze_registry, format_report
from platools.doctor.types import DoctorReport
from platools.types import ToolDef


def _load_registry(module_path: str | None) -> ToolRegistry:
    """Build a unified `ToolRegistry` from a user-provided module path.

    SDK consumers structure their app like this:

        # my_app/tools.py
        from platools import Platools
        platools = Platools()

        @platools.tool()
        def list_orders(): ...

    `platools doctor my_app.tools` imports that module and scans its
    top-level attributes for `Platools` instances; the registries on
    each instance are merged into a fresh composite registry the
    analyzer can walk. No `Platools` instances → empty registry, which
    surfaces as "Tools: 0 registered" in the report.
    """
    from platools import Platools

    composite = ToolRegistry()
    if not module_path:
        return composite

    module = importlib.import_module(module_path)
    for attr_name in dir(module):
        attr = getattr(module, attr_name, None)
        if isinstance(attr, Platools):
            for tool in attr.registry.all():
                if tool.name not in composite:
                    composite.register(tool)
    return composite


def _report_to_json(report: DoctorReport) -> str:
    return json.dumps(
        {
            "tool_count": report.tool_count,
            "errors": [
                {
                    "code": f.code,
                    "message": f.message,
                    "tool": f.tool,
                    "param": f.param,
                }
                for f in report.errors()
            ],
            "warnings": [
                {
                    "code": f.code,
                    "message": f.message,
                    "tool": f.tool,
                    "param": f.param,
                }
                for f in report.warnings()
            ],
            "info": [
                {
                    "code": f.code,
                    "message": f.message,
                    "tool": f.tool,
                    "param": f.param,
                }
                for f in report.infos()
            ],
        },
        indent=2,
    )


def run_doctor(
    *,
    module_path: str | None = None,
    output_json: bool = False,
    out: TextIO | None = None,
) -> int:
    """Programmatic entry point — used by both the CLI and the test suite.

    Returns the process exit code (0 = no errors, 1 = errors present).
    `out` is a text-mode file-like object; defaults to `sys.stdout` so
    the CLI behaves naturally and tests can pass `io.StringIO()` to
    capture output without monkey-patching.
    """
    sink: TextIO = out if out is not None else sys.stdout
    registry = _load_registry(module_path)
    report = analyze_registry(registry)

    if output_json:
        sink.write(_report_to_json(report) + "\n")
    else:
        sink.write(format_report(report))
    return 1 if report.has_errors() else 0


def doctor_command(argv: Sequence[str] | None = None) -> int:
    """`platools doctor` argparse front-end.

    Stdlib argparse instead of click so the SDK doesn't take a third-
    party CLI dep. The flag surface is small enough that argparse is
    fine and the wheel stays light.
    """
    parser = argparse.ArgumentParser(
        prog="platools doctor",
        description="Static tool-graph analyzer for Platools (PLATOS-18).",
    )
    parser.add_argument(
        "module",
        nargs="?",
        help="Optional dotted module path to import before analysis "
        "(e.g. `my_app.tools`). Loads the @platools.tool() decorators "
        "by side-effect into the global registry.",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Emit a machine-readable JSON report instead of the text shape.",
    )
    args = parser.parse_args(argv)
    return run_doctor(module_path=args.module, output_json=args.json)


# Direct programmatic helper used by tests that want to feed an
# explicit list of tools without touching the global registry.
def run_doctor_on_tools(tools: list[ToolDef]) -> DoctorReport:
    """Analyze a pre-loaded list of `ToolDef` and return the report."""
    from platools.doctor import analyze_tools

    return analyze_tools(tools)
